**Neural Network-Based Classification Model: Iris Flower Species Prediction**

In this project, a neural network classification model is developed to predict the species of flowers based on the well-known **Iris dataset**. The dataset consists of measurements of **sepal length, sepal width, petal length, and petal width** for three different species of Iris flowers: **Setosa, Versicolor, and Virginica**.

Since this is a **multiclass classification problem**, the model is designed with a **Softmax activation function** in the output layer, allowing it to assign probabilities to each class and make accurate predictions.

🔍 **More details and comments are available at the end of the file.**

**#0 Imports**

In [1]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

**#1 Data loading**

In [2]:
data = load_iris()
print(data.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [3]:
print(data['data'][:5])  
print(data['feature_names'])  
print(data['target'][:5]) 
print(data['target_names']) 

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
[0 0 0 0 0]
['setosa' 'versicolor' 'virginica']


In [4]:
X = data['data'] 
y = data['target']

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [6]:
print(X_train.shape, X_test.shape)

(120, 4) (30, 4)


In [7]:
# Standardization
scaler = StandardScaler()
X_train_standardized = scaler.fit_transform(X_train)
X_test_standardized = scaler.transform(X_test)

**#2 Model declaration**

In [8]:
model = Sequential([
    Dense(64, activation='relu', input_dim=4),  # 4 input features
    Dense(32, activation='relu'),
    Dense(3, activation='softmax')   # 3 classes → softmax
])

C:\Users\Mateusz\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


**#3 Model compilation**

In [9]:
adam_opt = tf.keras.optimizers.Adam(learning_rate=0.001)

In [10]:
model.compile(
    optimizer=adam_opt,
    loss="sparse_categorical_crossentropy",  # Because labels are numbers (0,1,2)
    metrics=["accuracy"]
)

**#4 Model preview**

In [11]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 64)                  │             320 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 3)                   │              99 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,499 (9.76 KB)

 Trainable params: 2,499 (9.76 KB)

 Non-trainable params: 0 (0.00 B)

**#5 Training the model**

In [12]:
history = model.fit(
    X_train_standardized, 
    y_train, 
    epochs=50, 
    validation_split=0.2  # 20% of training data for validation
)

Epoch 1/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 2s 161ms/step - accuracy: 0.6471 - loss: 0.9803 - val_accuracy: 0.6667 - val_loss: 0.9468
Epoch 2/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - accuracy: 0.7682 - loss: 0.9244 - val_accuracy: 0.7500 - val_loss: 0.8851
Epoch 3/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - accuracy: 0.7995 - loss: 0.8544 - val_accuracy: 0.7500 - val_loss: 0.8300
Epoch 4/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - accuracy: 0.8698 - loss: 0.7712 - val_accuracy: 0.7500 - val_loss: 0.7783
Epoch 5/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step - accuracy: 0.8346 - loss: 0.7316 - val_accuracy: 0.7500 - val_loss: 0.7291
Epoch 6/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - accuracy: 0.8307 - loss: 0.6879 - val_accuracy: 0.7500 - val_loss: 0.6828
Epoch 7/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 56ms/step - accuracy: 0.8568 - loss: 0.6125 - val_accuracy: 0.8333 - val_loss: 0.6400
Epoch 8/50
3/3 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - accuracy: 0.8320 - loss: 0.5853 - val_accuracy: 0.8333 - val_loss: 0.5998

**#6 Model evaluation**

In [13]:
test_loss, test_acc = model.evaluate(X_test_standardized, y_test)
print(f"Accuracy on the test set: {test_acc:.2f}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step - accuracy: 0.9333 - loss: 0.1742
Accuracy on the test set: 0.93


**#7 Summary**

In conclusion, the accuracy of 87% is a good value for Iris classification, which means that the model performs well in distinguishing flower species based on their characteristics.